src/main/java/de/hysky/skyblocker/skyblock/end/TheEnd.java

```diff
@@ -0,0 +1,232 @@
+            }
+
+
+        });
+        // Reset when changing island
+        // TODO: Replace when a changed island event is added
+        ClientPlayConnectionEvents.JOIN.register((handler, sender, client) -> {
+            resetLocation();
+            save();
+            load();
+        });
+        // Save when leaving as well
+        ClientLifecycleEvents.CLIENT_STOPPING.register((client) -> save());
+
+        ClientReceiveMessageEvents.GAME.register((message, overlay) -> {
+            String lowerCase = message.getString().toLowerCase();
+            if (lowerCase.contains("tremor") && stage != 0) stage+=1; // TODO: If stage is 0 re-scan.
+            else if (lowerCase.contains("rises from below")) stage = 5;
+            else if (lowerCase.contains("protector down") || lowerCase.contains("has risen")) resetLocation();
+            else return;
+            EndHudWidget.INSTANCE.update();
+        });
+
+        WorldRenderEvents.AFTER_TRANSLUCENT.register(TheEnd::renderWaypoint);
+    }
+
+    private static void resetLocation() {
+        stage = 0;
+        currentProtectorLocation = null;
+    }
+
+    public static void onEntityDeath(Entity entity) {
+        if (!(entity instanceof EndermanEntity enderman) || !isZealot(enderman)) return;
+        if (hitZealots.contains(enderman.getUuid())) {
+            //MinecraftClient.getInstance().player.sendMessage(Text.literal("You killed a zealot!!!"));
+            if (isSpecialZealot(enderman)) {
+                zealotsSinceLastEye = 0;
+                eyes++;
+            }
+            else zealotsSinceLastEye++;
+            zealotsKilled++;
+            dirty = true;
+            hitZealots.remove(enderman.getUuid());
+            EndHudWidget.INSTANCE.update();
+
+        }
+
+    }
+
+    public static boolean isZealot(EndermanEntity enderman) {
+        if (enderman.getName().getString().toLowerCase().contains("zealot")) return true; // Future-proof. If they someday decide to actually rename the entities
+        assert MinecraftClient.getInstance().world != null;
+        List<ArmorStandEntity> entities = MinecraftClient.getInstance().world.getEntitiesByClass(
+                ArmorStandEntity.class,
+                enderman.getDimensions(null).getBoxAt(enderman.getPos()).expand(1),
+                armorStandEntity -> armorStandEntity.getName().getString().toLowerCase().contains("zealot"));
+        if (entities.isEmpty()) {
+            return false;
+        }
+        return entities.get(0).getName().getString().toLowerCase().contains("zealot");
+    }
+
+    public static boolean isSpecialZealot(EndermanEntity enderman) {
+        return isZealot(enderman) && enderman.getCarriedBlock() != null && enderman.getCarriedBlock().isOf(Blocks.END_PORTAL_FRAME);
+    }
+
+    /**
+     * Loads if needed
+     */
+    public static void load() {
```

Comment:
Instead I suggest having one file in the main config folder like the museum item cache (maybe you could put it in the profile stats storage folder?) which contains the data for all profiles - probably much easier to save to & load from as you could just serialize the entire map to json.

Reply:
changed it to one file, not quite sure if I did it right